File: typescript/packages/node/src/core/disk/stat.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { DiskAccessor } from '../../accessor/disk.ts'
import { stat as fsStat } from 'node:fs/promises'
import path from 'node:path'
import { enoent, FileStat, FileType, guessType, type PathSpec } from '@struktoai/mirage-core'
import { resolveSafe } from './utils.ts'

export async function stat(accessor: DiskAccessor, p: PathSpec): Promise<FileStat> {
  const virtual = p.stripPrefix
  const full = resolveSafe(accessor.root, virtual)
  let st
  try {
    st = await fsStat(full)
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === 'ENOENT') {
      throw enoent(p)
    }
    throw err
  }
  const modified = st.mtime.toISOString()
  const name = path.basename(full)
  if (st.isDirectory()) {
    return new FileStat({
      name,
      size: null,
      modified,
      type: FileType.DIRECTORY,
    })
  }
  return new FileStat({
    name,
    size: st.size,
    modified,
    fingerprint: modified,
    type: guessType(name),
  })
}

File: python/mirage/core/disk/stat.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from datetime import datetime, timezone
from pathlib import Path

import aiofiles.os
from aiofiles.os import path as aio_path

from mirage.accessor.disk import DiskAccessor
from mirage.cache.index import IndexCacheStore
from mirage.core.timeutil import to_iso_z
from mirage.types import FileStat, FileType, PathSpec
from mirage.utils.filetype import guess_type


def _resolve(root: Path, path: str) -> Path:
    relative = path.lstrip("/")
    resolved = (root / relative).resolve()
    resolved.relative_to(root)
    return resolved


async def stat(accessor: DiskAccessor,
               path: PathSpec,
               index: IndexCacheStore = None) -> FileStat:
    if isinstance(path, str):
        path = PathSpec(original=path, directory=path)
    if isinstance(path, PathSpec):
        prefix = path.prefix
        path = path.original
    virtual = path
    if prefix and path.startswith(prefix):
        rest = path[len(prefix):]
        if prefix.endswith("/") or rest == "" or rest.startswith("/"):
            path = rest or "/"
    root = accessor.root
    p = _resolve(root, path)
    if not await aio_path.exists(p):
        raise FileNotFoundError(virtual)
    st = await aiofiles.os.stat(p)
    modified = to_iso_z(datetime.fromtimestamp(st.st_mtime, tz=timezone.utc))
    if await aio_path.isdir(p):
        return FileStat(name=p.name,
                        size=None,
                        modified=modified,
                        type=FileType.DIRECTORY)
    return FileStat(name=p.name,
                    size=st.st_size,
                    modified=modified,
                    fingerprint=modified,
                    type=guess_type(p.name))
